### ISSUE TITLE
geometry instancing problem

### ISSUE THREAD

**Issue by lostscience** [UPBGE/blender#145]
10,000 cubes with geometry instancing and a red material causes a crash in the upbge.

**Comment by panzergame** [comment]
Something like that : http://www.pasteall.org/blend/42633 ?

**Comment by lostscience** [comment]
With instancing the framerate is about ten.With just regular cubes in the
scene it is forty or sixty frames.I did not save the blend with regular
cubes in it and the bge crashed.

**Comment by DCubix** [comment]
Why was this closed?

**Comment by Wisaam** [comment]
Maybe you're duplicating them using Shift+D instead of Alt+D.
And use the same material for all of them.

**Comment by panzergame** [comment]
the crash is fixed in 7dde1096f6ca223a95fda35d547c05a496858d06. About performance issue if you don't gives infos about your GPU i can't do anything, also as Wisaam said, are you duplicating object with shared mesh ?

**Comment by lostscience** [comment]
I did not try alt d yet.

**Comment by lostscience** [comment]
I was using array modifier in object mode and seperate by loose part in edit mode.

**Comment by panzergame** [comment]
@lostscience : Did you know use alt+d ?

**Comment by lostscience** [comment]
I did use alt+d. 
For 1200 blocks with geometry instancing i get a framerate of 30.On a hewlletpackard desktop pc.
![wsacapture](https://cloud.githubusercontent.com/assets/16541304/16714661/c1004626-468f-11e6-9ce9-6718a0cc4727.PNG)
![screenshot 2016-07-10 11 05 12](https://cloud.githubusercontent.com/assets/16541304/16714700/d9ba1e3e-4690-11e6-87a2-70e5cc8f3ab8.png)

**Comment by youle31** [comment]
maybe answers can be found there: https://blenderartists.org/forum/showthread.php?367081-GLSL-Grass-Geometry-Shader-update-v-2-2-AMD-now-supported!

**Comment by lostscience** [comment]
The blend crashes the upbge.

**Comment by youle31** [comment]
@lostscience : could you start standalone player and show us the infos messages in the console please? http://www.pasteall.org/pic/show.php?id=105275

**Comment by lostscience** [comment]
How would i do that?Could you explain?

**Comment by Akira1San** [comment]
Toggle system console. In the render tab -> Stand alone player =  press the start button. :)

**Comment by lostscience** [comment]
I figured it out.
This what i have with the standalone player with vertex buffer objects and vsynch on.
That is with block easy demo.And i get a framerate 30.

```
Read new prefs: C:\Users\barbara\AppData\Roaming\Blender Foundation\Blender\2.77\config\userpref.blend
AL lib: (EE) UpdateDeviceParams: Failed to set 44100hz, got 48000hz instead
found bundled python: C:\Users\barbara\Desktop\UPBGEv0.0.9b2.77Windows64\2.77\python
*****************
Edge Roundifier
*****************
Imported multifiles
read blend: C:\Users\barbara\Desktop\blockWorldEasyDemo249x0x\blockWorldEasyDemo249x2x.blend
AL lib: (EE) UpdateDeviceParams: Failed to set 44100hz, got 48000hz instead
found bundled python: C:\Users\barbara\Desktop\UPBGEv0.0.9b2.77Windows64\2.77\python
GL_VENDOR: ATI Technologies Inc.
GL_RENDERER: AMD Radeon HD 8200 / R3 Series
GL_VERSION:  4.5.13431 Compatibility Profile Context 16.150.2211.0
Supported Extensions...
 GL_ARB_shader_objects supported?       yes.
 GL_ARB_vertex_shader supported?        yes.
 ----------Details----------
  Max uniform components.16384
  Max varying floats.128
  Max vertex texture units.32
  Max combined texture units.192

 GL_ARB_fragment_shader supported?      yes.
 ----------Details----------
  Max uniform components.16384

 GL_ARB_texture_cube_map supported?     yes.
 ----------Details----------
  Max cubemap size.16384

 GL_ARB_multitexture supported?         yes.
 ----------Details----------
  Max texture units available.  8

 GL_ARB_texture_env_combine supported?  yes.
 GL_ARB_texture_non_power_of_two supported?  yes.
 GL_ARB_draw_instanced supported?  yes.

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: Player.001

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass

Hit object: None

Hit object: grass
Error Totblock: 1

```

**Comment by youle31** [comment]
@panzergame : hi,  I can reproduce the performance problem with huge amount of cubes: http://www.pasteall.org/blend/43068
(all cubes share the same material so you just have to check geometry instancing on one cube to see the difference). This happens both in VA and VBO

**Comment by panzergame** [comment]
@youle31 : Your test file is a bit erroneous, GI need to share as mush as possible meshes but is not the case in your file. And if we compare GI vs regular render for only one unique mesh GI a bit slower.

**Comment by youle31** [comment]
Ah!! ok :) So Lostscience, you've got your answer. To use geometry instancing, the objects must share the same mesh...

@panzergame : Do you think it would be possible to add an UI just below the checkbox geometry instancing to choose the mesh we want to use? I think like that it would be more clear that geometry instancing is only efficient for objects that use the same mesh... I'll maybe try to implement it if you are ok.

@lostscience : Can I close the issue?

**Comment by lostscience** [comment]
You can close issue.I did not understand that at first.

**Comment by lostscience** [comment]
I get 9.9 framerate for this file.Take a look and see if it is normal. https://www.sendspace.com/file/52ctd7

**Comment by youle31** [comment]
I don't know: there are more than 100 .blend. I don't have the time to check such an example. If you want to test geometry instancing: http://www.pasteall.org/blend/43093 (just check geometry instancing on 1 cube and see the framerate difference. I duplicated the cubes with ALT+D so the cubes share the same mesh. I have 1200 fps without geometry instancing in last upbge sources and 1800 fps with geometry instancing. We did all what we can do to try to help you, so I close the thread.
